src-electron/db/query-bitmap.js

```diff
@@ -68,6 +68,62 @@ WHERE (DATA_TYPE.NAME = ? OR DATA_TYPE.NAME = ?) AND DATA_TYPE.PACKAGE_REF IN ($
     .then(dbMapping.map.bitmap)
 }
 
+/**
+ * Select a bitmap matched by name and clusterId.
+ * @param {*} db
+ * @param {*} name
+ * @param {*} clusterId
+ * @param {*} packageIds
+ * @returns bitmap information or undefined
+ */
+async function selectBitmapByNameAndClusterId(db, name, clusterId, packageIds) {
+  let res = await dbApi
+    .dbAll(
+      db,
+      `
+    SELECT
+      BITMAP.BITMAP_ID,
+      DATA_TYPE.NAME AS NAME,
+      BITMAP.SIZE AS SIZE
+    FROM BITMAP
+    INNER JOIN DATA_TYPE ON BITMAP.BITMAP_ID = DATA_TYPE.DATA_TYPE_ID
+    WHERE (DATA_TYPE.NAME = ? OR DATA_TYPE.NAME = ?) AND DATA_TYPE.PACKAGE_REF IN (${dbApi.toInClause(
+      packageIds
+    )})`,
+      [name, name.toLowerCase()]
+    )
+    .then((rows) => rows.map(dbMapping.map.bitmap))
+
+  if (res && res.length == 1) {
+    return res[0]
+  } else {
+    return dbApi
+      .dbGet(
+        db,
+        `
+SELECT
```

Comment:
The query duplication here seems a bit sub-optimal, in that we have to remember to change both queries in sync.

If there is a good reason (you mentioned performance optimization) to not just do the one query that includes the cluster id, then maybe we could have a function that returns the query string and takes an argument for whether to include the cluster id check, and does string concatenation or some other mechanism to include it or not while keeping the rest of the query as a single string?  Similar for other places where we are adding such queries.

Reply:
Removed duplication